### DISCUSSION TITLE
How to add new plugins to a scene that has been imported.

### DISCUSSION THREAD

**Discussion by CGCooke** [mitsuba-renderer/mitsuba3#583] [Q&A] [answered]
Hello,

I'm following this [tutorial](https://mitsuba.readthedocs.io/en/stable/src/rendering/editing_a_scene.html).

How would I add new plugins (Geometry/Volumes etc) after the scene has been loaded?

For example, I would like to do something like this:

`
scene = mi.load_file("../scenes/simple.xml")

env_light = {
           'type':'envmap',
            'filename': 'abandoned_tank_farm_04.exr',
            'to_world': mi.ScalarTransform4f.rotate(axis=(1, 0, 0), angle=90),
}

params = mi.traverse(scene)
params['env_light'] = env_light
params.update()
`

However, when I try this, I get:

> ---------------------------------------------------------------------------
> KeyError                                  Traceback (most recent call last)
> Input In [5], in <cell line: 1>()
> ----> 1 params[' env_light'] = light
> 
> File /opt/anaconda3/envs/Mitsuba-Experiments/lib/python3.10/site-packages/mitsuba/python/util.py:60, in SceneParameters.__setitem__(self, key, value)
>      59 def __setitem__(self, key: str, value):
> ---> 60     cur, value_type, node, _ = self.properties[key]
>      62     cur_value = cur
>      63     if value_type is not None:
> 
> KeyError: ' env_light'

1. Can new plugins be added after a scene has been loaded?
2. If so, what is the best way to achieve this?

**Comment by njroussel** [comment, marked as answer]
Hi @CGCooke 

1. No, this is not currently supported. Once a scene has been loaded, only its parameters can be modified. No change of plugin types, or removal/addition of new plugins is possible.

2. The only workaround is to reload a scene with the new desired description. Mixing Python dictionaries with XML descriptions is possible but limited. For example, you can load individual plugins (not an entire scene) through `mi.load_string`, `mi.load_file` or `mi.load_dict` and then combine these pre-built objects into a scene or another plugin with `mi.load_dict`. Like so:
```python
import mitsuba as mi
mi.set_variant('llvm_ad_rgb')
bsdf = mi.load_string('<bsdf version="3.0.0" type="diffuse"/>')
cube = mi.load_dict({'type': 'cube', 'bsdf': bsdf})
```
The general recommendation is to commit to one format, either XML or Python dictionnaries.

**Reply by CGCooke** [reply]
Thank-you @njroussel for a thorough answer, very clear.
